langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java

```diff
@@ -0,0 +1,506 @@
+import com.oracle.coherence.ai.Vector;
+import com.oracle.coherence.ai.VectorIndexExtractor;
+
+import com.oracle.coherence.ai.search.SimilaritySearch;
+import com.oracle.coherence.common.base.Logger;
+
+import com.tangosol.internal.util.processor.CacheProcessors;
+import com.tangosol.net.Coherence;
+import com.tangosol.net.NamedMap;
+import com.tangosol.net.Session;
+
+import com.tangosol.util.Filter;
+import com.tangosol.util.UUID;
+
+import com.tangosol.util.ValueExtractor;
+import dev.langchain4j.data.document.Metadata;
+
+import dev.langchain4j.data.embedding.Embedding;
+
+import dev.langchain4j.data.segment.TextSegment;
+
+import dev.langchain4j.store.embedding.EmbeddingMatch;
+import dev.langchain4j.store.embedding.EmbeddingSearchRequest;
+import dev.langchain4j.store.embedding.EmbeddingSearchResult;
+import dev.langchain4j.store.embedding.EmbeddingStore;
+import dev.langchain4j.store.embedding.RelevanceScore;
+
+import java.io.Serializable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import java.util.function.ToDoubleFunction;
+
+import java.util.stream.Collectors;
+
+import static dev.langchain4j.internal.Utils.isNullOrEmpty;
+import static dev.langchain4j.internal.ValidationUtils.ensureTrue;
+
+/**
+ * An {@link EmbeddingStore} backed by an Oracle Coherence {@link NamedMap}.
+ */
+public class CoherenceEmbeddingStore
+        implements EmbeddingStore<TextSegment> {
+    /**
+     * A {@link Comparator} to sort an {@link EmbeddingMatch} based on score, where the
+     * highest score is first.
+     */
+    private static final Comparator<EmbeddingMatch<?>> COMPARATOR = reverseComparingDouble(EmbeddingMatch::score);
+
+    /**
+     * The {@link ValueExtractor} to extract the float vector from a {@link DocumentChunk}.
+     */
+    private static final ValueExtractor<DocumentChunk, Vector<float[]>> EXTRACTOR = ValueExtractor.of(DocumentChunk::vector);
+    
+    /**
+     * The default {@link NamedMap} name.
+     */
+    public static final String DEFAULT_MAP_NAME = "documentChunks";
+
+    /**
+     * The {@link NamedMap} used to store the {@link DocumentChunk document chunks}.
+     */
+    protected final NamedMap<DocumentChunk.Id, DocumentChunk> documentChunks;
+
+    protected final boolean normalizeEmbeddings;
+
+    /**
+     * Create an {@link CoherenceEmbeddingStore}.
+     * <p>
+     * This method is protected, instances of {@link CoherenceEmbeddingStore}
+     * are created using the builder.
+     *
+     * @param namedMap             the {@link NamedMap} to contain the {@link DocumentChunk document chunks}
+     * @param normalizeEmbeddings  {@code true} if this {@link CoherenceEmbeddingStore} should call
+     *                             {@link Embedding#normalize()} on embeddings when adding or searching
```

Comment:
```suggestion
     *                             {@link Embedding#normalize()} on embeddings when adding them to the store
```
It does not seeem that it happens when searching, also it does not seem to be required on search.

Please fix it everywhere in this class (5 times)

Reply:
We actually do normalize query embedding if this field is set (see L161-163), and I do think it's required to ensure that we are comparing apples to apples (ie. if the embeddings in the store are normalized, we need to normalize search vector/embedding as well).

So the doc above is correct, unless we change the behavior.

Reply:
Sorry, I misunderstood it